src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/math/SynchronizedUnivariateSolver.java

```diff
@@ -0,0 +1,385 @@
+package org.broadinstitute.hellbender.tools.coveragemodel.math;
+
+import org.apache.commons.math3.analysis.solvers.AbstractUnivariateSolver;
+import org.apache.commons.math3.exception.NoBracketingException;
+import org.apache.commons.math3.exception.TooManyEvaluationsException;
+import org.apache.commons.math3.util.FastMath;
+import org.broadinstitute.hdf5.Utils;
+import org.broadinstitute.hellbender.utils.param.ParamUtils;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * This class implements a synchronized univariate solver for solving multiple independent equations.
+ * It is to be used in situations where function queries have a costly overhead, though, simultaneous
+ * queries of multiple functions have the same overhead as single queries.
+ *
+ * Consider the tasking of solving N independent equations:
+ *
+ *      f_1(x_1) = 0,
+ *      f_2(x_2) = 0,
+ *      ...
+ *      f_N(x_N) = 0
+ *
+ * One approach is to solve these equations sequentially. In certain situations, each function
+ * evaluation may be cheap but could entail a costly overhead (e.g. if the functions are evaluated
+ * in a distributed architecture). It is desirable to minimize this overhead by bundling as many
+ * function calls as possible, and querying the function in "chunks".
+ *
+ * Consider te ideal situation where function evaluations are infinitely cheap, however, each query has
+ * a considerable overhead time of \tau. Also, let us assume that the overhead of simultaneously
+ * querying {f_1(x_1), ..., f_N(x_N)} is the same as that of a single query, i.e. f_i(x_i). If the
+ * univariate solver requires k queries on average, the overhead cost of the sequential approach is
+ * O(k N \tau). By making simultaneous queries, this class reduces the overhead to O(k \tau).
+ *
+ * This is achieved by instantiating N threads for the N univariate solvers, accumulating their queries
+ * and suspending them until all threads announce their required query.
+ *
+ * TODO github/gatk-protected issue #853
+ * @implNote In the current implementation, we make a thread for each solver. This is fine if the number
+ * of equations is reasonably small (< 200). In the future, the class must take a max number
+ * of threads and limit concurrency.
+ *
+ * @author Mehrtash Babadi &lt;mehrtash@broadinstitute.org&gt;
+ */
+public final class SynchronizedUnivariateSolver {
+
+    /**
+     * Default value for the absolute accuracy of function evaluations
+     */
+    private static final double DEFAULT_FUNCTION_ACCURACY = 1e-15;
+
+    private enum SolverClass {
```

Comment:
As discussed in person, I'm inclined to let this class go through as long as we file an issue to refactor later.